supply-chain/src/lib.rs

```diff
@@ -0,0 +1,595 @@
+}
+
+async fn transfer_nft(nft_program_id: ActorId, to: ActorId, token_id: U256) {
+    msg::send_and_wait_for_reply::<NFTEvent, _>(
+        nft_program_id,
+        NFTAction::Transfer { to, token_id },
+        0,
+    )
+    .unwrap()
+    .await
+    .expect("Unable to decode NFTEvent");
+}
+
+async fn receive(ft_program_id: ActorId, seller: ActorId, item: &Item) {
+    let elapsed_time = exec::block_timestamp() - item.shipping_time;
+    // By default, all tokens transferred to a seller,
+    let (mut to, mut amount) = (seller, item.price);
+
+    // but if a seller spent more time than it was agreed...
+    if elapsed_time > item.delivery_time {
+        // ...and extremely late (more than or exactly 2 times in this example),
+        if elapsed_time >= item.delivery_time * 2 {
+            // then all tokens refunded to a buyer...
+            to = msg::source();
+        } else {
+            // ...or a half of tokens refunded to a buyer and...
+            transfer_tokens(
+                ft_program_id,
+                exec::program_id(),
+                msg::source(),
+                (item.price + 1) / 2,
+            )
+            .await;
+
+            // ...another half transferred to a seller.
+            amount /= 2;
+        }
+    }
+
+    transfer_tokens(ft_program_id, exec::program_id(), to, amount).await;
+}
+
+#[derive(Default)]
+struct SupplyChain {
+    items: BTreeMap<U256, Item>,
+
+    producers: BTreeSet<ActorId>,
+    distributors: BTreeSet<ActorId>,
+    retailers: BTreeSet<ActorId>,
+
+    ft_program_id: ActorId,
+    nft_program_id: ActorId,
+}
+
+impl SupplyChain {
+    fn check_producer(&self) {
+        if !self.producers.contains(&msg::source()) {
+            panic!("msg::source() must be a producer");
+        }
+    }
+
+    fn check_distributor(&self) {
+        if !self.distributors.contains(&msg::source()) {
+            panic!("msg::source() must be a distributor");
+        }
+    }
+
+    fn check_retailer(&self) {
+        if !self.retailers.contains(&msg::source()) {
+            panic!("msg::source() must be a retailer");
+        }
+    }
+
+    /// Produces one item with a name and notes and replies with its ID.
+    /// Transfers created NFT for an item to a producer.
+    ///
+    /// Requirements:
+    /// * `msg::source()` must be a producer in a supply chain.
+    ///
+    /// Arguments:
+    /// * `name`: an item's name.
+    /// * `notes`: an item's notes.
+    async fn produce_item(&mut self, name: String, notes: String) {
+        self.check_producer();
+
+        // After minting NFT for an item,
+        // an item gets an ID equal to the ID of its NFT.
+        let item_id = match msg::send_and_wait_for_reply(self.nft_program_id, NFTAction::Mint, 0)
+            .unwrap()
+            .await
+            .expect("Unable to decode NFTEvent") {
+                NFTEvent::Transfer { from, to, token_id } if from == ActorId::default() && to == exec::program_id() => token_id,
```

Comment:
https://github.com/gear-tech/apps/blob/33a66f0cd9ac5a93f9894256b3b73fbc28faa55d/nft-example/src/lib.rs#L45
`ActorId::default()` equals `ZERO_ID`.
`from` is `ZERO_ID` only after minting, so I decided to add this check to avoid unexpected behavior.

Reply:
I don't think it's necessary in this context because it doesn't matter who nft comes from. NFT contract can change source(instead it will send it from its ActorId) then your contract will break down.

Reply:
You only need to know that it has come to you